server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java

```diff
@@ -0,0 +1,482 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.approximate;
+
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.PointRangeQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.ScorerSupplier;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.DocIdSetBuilder;
+import org.apache.lucene.util.IntsRef;
+import org.opensearch.search.sort.SortOrder;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * An approximate-able version of {@link PointRangeQuery}. It creates an instance of {@link PointRangeQuery} but short-circuits the intersect logic
+ * after {@code size} is hit
+ */
```

Comment:
It might be worthwhile to highlight which parts are directly copied from `PointRangeQuery` and which parts are specific to the approximation.

Moving forward, we might be able to advocate for some refactoring to `PointRangeQuery` to reduce the need for duplication.

Reply:
will do, I would like to delegate as much as possible to `PointRangeQuery`, right now we re-use some stuff like `matches()`, `relate()`. These are the only things that we duplicate, of course we would need our own scorer and intersect implementation. Also, if the size of the tree is smaller than the size we want we just delegate to `PointRangeQuery`. Will add more comments to make things clearer.